src/marin/processing/classification/autoscaler.py

```diff
@@ -0,0 +1,456 @@
+        """Remove idle actors from the pool."""
+        with self.actor_task_metadata_lock:
+            removed = 0
+            actors_to_remove = []
+
+            for actor, futures in list(self.actor_futures.items()):
+                if removed >= count:
+                    break
+                if len(futures) == 0 and len(self.actors) > self.min_actors:
+                    actors_to_remove.append(actor)
+                    removed += 1
+
+            for actor in actors_to_remove:
+                self.actors.remove(actor)
+                del self.actor_futures[actor]
+                ray.kill(actor)
+
+            if removed > 0:
+                logger.info(f"Scaled down: removed {removed} actors. Pool size now: {len(self.actors)}")
+
+    def _get_least_loaded_actor(self):
+        """Get the actor with the least number of pending tasks."""
+        if not self.actors:
+            return None
+
+        logger.debug("Trying to dispatch the task")
+        with self.actor_task_metadata_lock:
+            alive_actors = []
+            for actor in self.actors:
+                ping_received = actor.ping.remote()
+                logger.debug(f"Pinging actor: {actor}")
+                ping_received_ready, _ = ray.wait([ping_received], num_returns=1, timeout=1.0)
+
+                logger.debug("Ping wait result", extra={"actor": str(actor), "ready": bool(ping_received_ready)})
+                if ping_received_ready:
+                    try:
+                        result = ray.get(ping_received_ready, timeout=5.0)
+                        if result:
+                            logger.debug(f"Ping received from actor: {actor}")
+                            alive_actors.append(actor)
+                    except Exception as e:
+                        logger.error(f"Error received trying to ping actor: {e}")
+
+            self.actors = alive_actors
+            alive_actor_futures = {}
+            for actor, futures in self.actor_futures.items():
+                alive_actor_futures[actor] = futures
+
+            self.actor_futures = alive_actor_futures
+            available_actors = [
+                actor
+                for actor in self.actors
+                if self.max_tasks_per_actor is None or len(self.actor_futures.get(actor, [])) < self.max_tasks_per_actor
+            ]
+
+        if not available_actors:
+            return None
+
+        return min(available_actors, key=lambda a: len(self.actor_futures.get(a, [])))
+
+    def _dispatch_task(self, task: list[dict[str, Any]]) -> ray.ObjectRef:
+        """Dispatch a single task to an available actor."""
+
+        actor = self._get_least_loaded_actor()
+        if actor is None:
+            raise RuntimeError("No actors available")
+
+        with self.actor_task_metadata_lock:
+            current_load = len(self.actor_futures.get(actor, []))
+            if self.max_tasks_per_actor is not None and current_load >= self.max_tasks_per_actor:
```

Comment:
why would we ever have more than one task per actor? actors serialize work anyway right? we could just ray.wait([pending_request_list]) in the queueing thread and schedule as soon as work finishes?

or are we worried about the latency?